### PR TITLE
Add type annotations, part 1

### PR DESCRIPTION
**Pull request by andlaus**
This is the kick-off PR of another long and winding quest; this time the goal is to add comprehensive type annotations to everything so that `mypy --disallow-untyped-defs` passes for odxtools without any complaints. Note that this quest will include many side-tasks; I indent to put them into separate PRs as much as possible, but it is about as certain as death and taxes that I'll occasionally miss a few of these "drive-by fixes"...

Andreas Lauser &lt;andreas.lauser@mercedes-benz.com&gt;, on behalf of [MBition GmbH](https://mbition.io/).
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

File: odxtools/cli/dummy_sub_parser.py [hunk 1]
```diff
@@ -14,19 +15,19 @@ class DummyTool:
     should bail out.
     """
 
-    def __init__(self, tool_name, error):
+    def __init__(self, tool_name: str, error: Exception):
         self._odxtools_tool_name_ = tool_name
         self._error = error
 
-    def add_subparser(self, subparser_list):
+    def add_subparser(self, subparser_list: argparse._SubParsersAction) -> None:
```
Comment (kayoub5) [inline odxtools/cli/dummy_sub_parser.py:22]: `_SubParsersAction` is private

File: odxtools/isotp_state_machine.py [hunk 2]
```diff
@@ -28,29 +29,29 @@ class IsoTpStateMachine:
...
 
-        E.g., add some data to a telegram, etc.
+        E.g., add some data to a telegram, etc. Returns a generator of
+        (receive_id, payload_data) tuples.
         """
         try:
             telegram_idx = self._can_rx_ids.index(rx_id)
         except ValueError:
-            return False  # unknown CAN ID
+            return  # unknown CAN ID
```
Comment (kayoub5) [inline odxtools/isotp_state_machine.py:52]: behavior change?

Comment (andlaus) [reply odxtools/isotp_state_machine.py:52]: not really: this method was always called with `do_yield = True`, and I'm honestly not sure anymore why I've implemented that option in the first place...

File: odxtools/cli/dummy_sub_parser.py (hunk 1 above)
Comment (andlaus) [reply odxtools/cli/dummy_sub_parser.py:22]: yes, but how to annotate this correctly? `Any`?

Comment (kayoub5) [reply odxtools/cli/dummy_sub_parser.py:22]: ## Option 1
https://github.com/python/typeshed/issues/7539#issuecomment-1076581049

## Option 2

Even if python makes changes and `_SubParsersAction` is no longer there, only mypy will report it, but the code itself will still work.

```suggestion
    def add_subparser(self, subparser_list: "argparse._SubParsersAction") -> None:
```

Comment (andlaus) [reply odxtools/cli/dummy_sub_parser.py]: cool. I picked option two: e8009c70294dc4c2b10c95a9011da0e3e6fe9d48

File: odxtools/cli/dummy_sub_parser.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: MIT
 import argparse
 import sys
+from typing import List
```
Comment (kayoub5) [inline odxtools/cli/dummy_sub_parser.py:4]: added but not used

Comment (kayoub5) [reply odxtools/cli/dummy_sub_parser.py:4]: other than this, LGTM

Comment (andlaus) [reply odxtools/cli/dummy_sub_parser.py]: okay, I've removed all unused imports (at least those flagged by `ruff`): 58c013ac695db3e3687d87e83e2228d8769637cf

**Pull request merged**